v-next/hardhat/src/internal/builtin-plugins/artifacts/artifact-manager.ts

```diff
@@ -0,0 +1,405 @@
+      }
+
+      this.#throwNotFoundError(
+        contractNameOrFullyQualifiedName,
+        bareNameToFullyQualifiedNameMap,
+      );
+    }
+
+    const fqns = bareNameToFullyQualifiedNameMap.get(
+      contractNameOrFullyQualifiedName,
+    );
+
+    if (fqns === undefined || fqns.size === 0) {
+      this.#throwNotFoundError(
+        contractNameOrFullyQualifiedName,
+        bareNameToFullyQualifiedNameMap,
+      );
+    }
+
+    if (fqns.size !== 1) {
+      throw new HardhatError(HardhatError.ERRORS.ARTIFACTS.MULTIPLE_FOUND, {
+        contractName: contractNameOrFullyQualifiedName,
+        candidates: Array.from(fqns).join(EOL),
+      });
+    }
+
+    const [fqn] = fqns;
+
+    return fqn;
+  }
+
+  #throwNotFoundError(
+    contractNameOrFullyQualifiedName: string,
+    bareNameToFullyQualifiedNameMap: Map<string, ReadonlySet<string>>,
+  ): never {
+    const similarNames = this.#getSimilarContractNames(
+      contractNameOrFullyQualifiedName,
+      bareNameToFullyQualifiedNameMap,
+    );
+
+    const suggestion = this.#formatSimilarNameSuggestions(
+      contractNameOrFullyQualifiedName,
+      similarNames,
+    );
+
+    throw new HardhatError(HardhatError.ERRORS.ARTIFACTS.NOT_FOUND, {
+      contractName: contractNameOrFullyQualifiedName,
+      suggestion,
+    });
+  }
+
+  #isFullyQualifiedName(name: string): boolean {
+    return name.includes(":");
+  }
+
+  #getFullyQualifiedNameFromArtifactAbsolutePath(artifactPath: string): string {
+    const relativePath = path.relative(this.#artifactsPath, artifactPath);
+    const sourceName = path.dirname(relativePath).split(path.sep).join("/");
+    const contractName = path.basename(relativePath, ".json");
+    return `${sourceName}:${contractName}`;
+  }
+
+  #getSimilarContractNames(
+    contractNameOrFullyQualifiedName: string,
+    bareNameToFullyQualifiedNameMap: Map<string, ReadonlySet<string>>,
+  ): string[] {
+    const names = this.#isFullyQualifiedName(contractNameOrFullyQualifiedName)
+      ? Array.from(bareNameToFullyQualifiedNameMap?.values() ?? [])
+          .map((s) => [...s])
+          .flat(1)
+      : Array.from(bareNameToFullyQualifiedNameMap?.keys() ?? []);
+
+    return names
+      .map(
+        (name) =>
+          [name, editDistance(name, contractNameOrFullyQualifiedName)] as const,
+      )
+      .sort(([_, d1], [__, d2]) => d1 - d2)
+      .filter(([_, d]) => d <= EDIT_DISTANCE_THRESHOLD)
+      .map(([name]) => name);
+  }
+
+  #formatSimilarNameSuggestions(
+    contractNameOrFullyQualifiedName: string,
+    similarNames: string[],
+  ): string {
+    switch (similarNames.length) {
+      case 0:
+        return "";
+      case 1:
+        return `Did you mean "${similarNames[0]}"?`;
+      default:
+        return `We found some that were similar:
+
+${similarNames.map((n) => `  * ${n}`).join(EOL)}
+
+Please replace "${contractNameOrFullyQualifiedName}" for the correct contract name wherever you are trying to read its artifact.
```

Comment:
Could we check if we got contract name or fully qualified name here and update the message accordingly? I think it's going to be easier to communicate with users that way.

Also, I don't think this error is exclusively related to reading artifacts so maybe it would be a good idea to make the messaging here more general.

Reply:
I changed it to reflect the contract name type.

I'd be ok with leaving "reading" as the action here. But open to better alternatives.

Reply:
> I changed it to reflect the contract name type.

Amazing! Thanks!

> I'd be ok with leaving "reading" as the action here. But open to better alternatives.

To be fair, now, reading it again, I think "read" makes sense here after all. Sorry for the noise.